README.md
#<img src="https://raw.githubusercontent.com/kaepora/miniLock/master/src/img/icon128.png" width="48"> miniLock  
##File encryption software that does more with less.

[<img src="https://travis-ci.org/kaepora/miniLock.svg?branch=master" width="80">](https://travis-ci.org/kaepora/miniLock)

* **[Code](https://github.com/kaepora/miniLock)** | **[Issues and Discussion](https://github.com/kaepora/miniLock/issues)**
* **HOPE X [Video](https://vimeo.com/101237413) | [Slides](https://minilock.io/files/HOPEX.pdf)**
* **Follow on [Twitter](https://twitter.com/minilockapp)** for latest project news

###Download
[<img src="https://minilock.io/img/chromeStore.png" width="250">](https://chrome.google.com/webstore/detail/minilock/mknafpnfcafklkflhenhaldedggpciao)

miniLock is currently audited, peer-reviewed software. An initial public release is available for Google Chrome and Chrome OS.  


###Software Status
miniLock was subjected to a cryptographic code audit carried out by [Cure53](https://cure53.de/) and with the support of the [Open Technology Fund](https://www.opentechfund.org/). Quoting from the conclusion of the audit report (**[PDF](https://minilock.io/files/cure53.pdf)**):

> Cure53 was tasked to test against the application security of miniLock and evaluate its cryptographic properties and promises. Over the course of four days of manual testing, no severe errors have been spotted. The code is soundly and neatly written, well structured, minimal and therefore offers no sinks for direct exploitation.

miniLock also ships with a Unit Test Kit located in `test`.

###0. Overview
miniLock is a small, portable file encryption software. The idea behind its design is that passphrase memorized by the user, along with their email address, can act as a complete, portable basis for a persistent public key identity and provide a full substitute for other key pair models, such as having the key pair stored on disk media (the PGP approach).  

Advancements in elliptic curve cryptography, specifically in systems such as `curve25519`, allow us to generate key pairs where the lengths of both public and private keys are relatively very small. This means that public keys become far easier to share (miniLock public keys, called *miniLock IDs*, fit inside less than half a tweet). This also means that a human-memorizable passphrase of adequate entropy can be used as the basis for deriving a private key.

When first opened, miniLock asks the user for their email address and a passphrase which it then uses to derive the user's private and public keys. Via this model, the user can establish their key pair on any computer that has miniLock installed using only this passphrase, without having to manage key files or identities and so on. Thanks to the small key sizes present in `curve25519`, we are guaranteed small, easily tweetable public keys and private keys that can be derived from passphrases. miniLock also contains checks to ensure the passphrases entered by the user are of sufficient entropy. miniLock will refuse weak passphrases completely and instead suggest stronger passphrases for use by the user.

miniLock then allows the user to encrypt files to other miniLock users via their miniLock IDs and decrypt files sent to them. miniLock's encryption format supports encrypting a single file to multiple recipients with a negligible increase in file size. Another feature is that analyzing a miniLock-encrypted file does not yield the miniLock IDs or identities of the sender or the recipient(s). Upon decryption, a legitimate recipient will be able to know and verify the identity of the sender, but will still be unable to determine the identity of other potential recipients.

miniLock file encryption provides both confidentiality and integrity. miniLock uses the [TweetNaCL](http://tweetnacl.cr.yp.to) cryptography library, [ported to JavaScript](https://github.com/dchest/tweetnacl-js), entirely due to its focus on simplicity, auditability and small size. Similarly, miniLock is designed to be as simple, portable, auditable and usable as possible. miniLock also uses [scrypt](http://www.tarsnap.com/scrypt.html) for "memory-hard" key derivation.

###1. User Flow
This section outlines an example user flow in order to help demonstrate how miniLock is supposed to help people.

Alice wants to send a scan of her passport to Bob. Sending it over email would compromise personal information, so Alice decided to first encrypt the scan using miniLock.

Bob opens miniLock and enters his email address and passphrase. miniLock displays his miniLock ID, which is tied to his passphrase and is persistent. He sends Alice his miniLock ID, which looks something like this:
`quBSaJLXKsRiaSrhgkPnswKocth711H29ZamMi1H9j4Mb`

Alice drags and drops her passport scan into miniLock and enters Bob's miniLock ID as the recipient. She clicks the encrypt button and sends the resulting `.minilock` file to Bob. Once Bob drags the encrypted file into miniLock, it automatically detects it as a miniLock-encrypted file destined to Bob, and decrypts and saves the passport scan on his computer.

###2. Key Derivation
miniLock uses the [zxcvbn](https://github.com/dropbox/zxcvbn) library in order to impose a strict limit on the amount of detected entropy present in entered passphrases. miniLock will not allow passphrases that fall below the threshold of 100 bits of entropy: if a passphrase of lower entropy is detected, miniLock will refuse it and will not allow access to encryption or decryption functions.

Users are encouraged to use passphrases which are easier to remember but harder to guess. If a user fails to enter a sufficiently entropic passphrase, miniLock will use a built-in dictionary of the 58,110 most common words in the English language to suggest a seven-word passphrase. This gives us a passphrase with approximately 111 bits of entropy, since 58110<sup>7</sup> ~= 2<sup>111</sup>.

Once we obtain a suitable passphrase, we hash it using `BLAKE2` and pass the resulting 32 bytes through `scrypt` in order to obtain the user's private `curve25519` key. `scrypt` is invoked using the following parameters::

* N = 2<sup>17</sup>
* r = 8,
* p = 1,
* L = 32

miniLock uses the email address entered by the user as the `scrypt` key derivation salt. Email addresses are unique by nature and therefore provide a good basis for a salt.

Once we obtain our 32-byte private key, the public key is derived for use with the TweetNaCL `curve25519-xsalsa20-poly1305` construction.

The user's `miniLock ID` consists of 33 bytes. The first 32 bytes are the user's `curve25519` public key. The last byte acts as a checksum: it is derived by hashing the first 32 bytes with `BLAKE2` set to a 1-byte output. After constructing the 33 bytes of the miniLock ID, it is encoded into a Base58 representation, meant to be easily communicable via email or instant messaging.


###3. File format
miniLock saves encrypted files as binary blobs with the following format:

```javascript
miniLock magic bytes (8 bytes)
Header length in bytes (4 bytes, little-endian)
Header bytes
Ciphertext bytes
```

miniLock magic bytes identify that this is a miniLock-encrypted file:

```javascript
0x6d, 0x69, 0x6e, 0x69,
0x4c, 0x6f, 0x63, 0x6b
```

The header itself is a stringified JSON object which contains information necessary for the recipients to decrypt the file. The JSON object has the following format:

```json
{
version: Version of the miniLock protocol used for this file (Currently 1) (Number)
ephemeral: Public key from ephemeral key pair used to encrypt decryptInfo object (Base64),
decryptInfo: {
	(One copy of the below object for every recipient)
	Unique nonce for decrypting this object (Base64): {
		senderID: Sender's miniLock ID (Base58),
		recipientID: miniLock ID of this recipient (used for verfication) (Base58),
		fileInfo: {
			fileKey: Key for file decryption (Base64),
			fileNonce: Nonce for file decryption (Base64),
			fileHash: BLAKE2 hash (32 bytes) of the ciphertext bytes. (Base64)
		} (fileInfo is encrypted to recipient's public key using long-term key pair) (Base64),
	} (encrypted to recipient's public key using ephemeral key pair) (Base64)
}
```

Note that the nonce used to encrypt `decryptInfo` is the same as the one used to encrypt `fileInfo`. Nonce reuse in this scenario is permitted since we are encrypting using different keys.

###4. File encryption
The sender begins by generating a new ephemeral `curve25519` key pair, `senderEphemeralSecret` and `senderEphemeralPublic`.

The sender's long-term keys are denoted as `senderSecret` and `senderPublic`.

A recipient `a`'s long-term keys are denoted as `recipientSecret[a]` and `recipientPublic[a]`.

The sender appends the bytes signalling the beginning of the header to the final encrypted file.

The file's filename is padded with 0x00 bytes until its length equals 256 bytes. The filename is then prepended to the plaintext prior to encryption. The filename is encrypted as its own 256-byte chunk (see chunking format below).

A random 32-byte `fileKey` and a random 16-byte `fileNonce` are generated and used to symmetrically encrypt the plaintext bytes using TweetNaCL's `xsalsa20-poly1305` construction. We encrypt the plaintext bytes by splitting the plaintext into 1048576-byte chunks. Each chunk is then encrypted using the following model:

```javascript
fullNonce0 = fileNonce || 0
encryptedChunk0 = length(chunk0) || nacl.secretbox(chunk0, fullNonce0, fileKey)
fullNonce1 = fileNonce || 1
encryptedChunk1 = length(chunk1) || nacl.secretbox(chunk1, fullNonce1, fileKey)
...
```

In the above example, the 24-byte `fullNonce` is acquired by concatenating the 16-byte `fileNonce` and 8-byte little-endian chunk number. Also,`length(chunk)` is a 4-byte little-endian plaintext chunk length.

The last chunk is encrypted as follows:
```javascript
fullNonceN = fileNonce || setMostSignificantBit(N)
encryptedChunkN = length(chunkN) || nacl.secretbox(chunkN, fullNonceN, fileKey)
```

The sender generates a `fileInfo` JSON object containing `fileKey`, `fileNonce` and `fileHash`. For every recipient `n`, the sender encrypts `fileInfo` using `senderSecret` and `recipientPublic[n]` and stores it within a `decryptInfo` object inside the JSON header along with `senderID`, as described in §3.

The name of the `decryptInfo` property in which the aforementioned elements are stored is a 24-byte nonce. The sender uses this nonce, along with `senderEphemeralSecret`, to encrypt the underlying JSON object asymmetrically to `recipientPublic[n]`, using TweetNaCL's `curve25519-xsalsa20-poly1305` construction. Note that this is done once for every recipient, creating a different `decryptInfo` object for every recipient, each labeled by their unique nonces.

Finally, the sender appends the bytes signalling the end of the header, followed by the ciphertext bytes.

TweetNaCL's `curve25519-xsalsa20-poly1305` construction provides authenticated encryption, guaranteeing both confidentiality and ciphertext integrity. The above header construction makes it impossible to determine the sender or recipient(s) of a miniLock-encrypted file simply by analyzing the ciphertext.

###5. File decryption
In order to decrypt the file, the recipient needs the information stored within the `decryptInfo` section of the header. They also will need the `ephemeral` property of the header in order to derive the shared secret, in conjunction with their long-term secret key, which can be used to decrypt their copy of the `decryptInfo` header object.

If there are multiple properties within `decryptInfo`, the recipient must iterate through every property until she obtains an authenticated decryption of the underlying object. Once a successful authenticated decryption of a `decryptInfo` property occurs, the recipient can then use the obtained `senderID` along with their long-term secret key to decrypt `fileKey` and use it in conjunction with `fileNonce` to perform an authenticated decryption of the ciphertext bytes.

Before or during file decryption, the recipient compares the 32-byte `BLAKE2` hash of the ciphertext against the decrypted `fileHash`. The recipient also compares the decrypted `recipientID` against their own miniLock ID. If any of these checks fail, decryption is aborted and an error is returned.

In order to decrypt the ciphertext bytes, the recipient breaks the ciphertext down to chunks consisting of the original 1048576 bytes of the plaintext chunk, plus the 4 bytes defining the chunk length and the 16 bytes defining the `poly1305` authentication code of that particular ciphertext chunk. Each chunk is then decrypted sequentially using the following model:

```javascript
fullNonce0 = fileNonce || 0
decryptedChunk0 = nacl.secretbox.open(chunk0, fullNonce0, fileKey)
fullNonce1 = fileNonce || 1
decryptedChunk1 = nacl.secretbox.open(chunk1, fullNonce1, fileKey)
...
```

After decryption, the recipient retrieves the filename from the first 256 bytes of the plaintext. The recipient strips the padding bytes from the filename and is now capable of saving the decrypted file.

If the authenticated asymmetric decryption of any header object fails, or the authenticated symmetric decryption of the file ciphertext fails, decryption is aborted and an error is returned.

###6. Key Identity Authentication
In PGP, public keys can be substantially larger than miniLock IDs, therefore necessitating the generation of key fingerprints which can then be used for out-of-band key identity authentication. With miniLock, users are able to authenticate out-of-band directly using the miniLock ID, due to its small length (approximately 45 Base58-encoded characters). Therefore, no specialized key identity authentication mechanism is required.

###7. Error Codes
miniLock will output these error codes when running into encryption or decryption errors. The user interface can then handle these errors in order to display information that is relevant to users:

**Encryption errors**
* `Error 1`: General encryption error

**Decryption errors**
* `Error 2`: General decryption error
* `Error 3`: Could not parse header
* `Error 4`: Invalid header version
* `Error 5`: Could not validate sender ID
* `Error 6`: File is not encrypted for this recipient
* `Error 7`: Could not validate ciphertext hash

###8. Caveats
miniLock is not intended to protect against malicious files being sent and received. It is the user's responsibility to vet the safety of the files they send or receive over miniLock. miniLock cannot protect against malware being sent over it.

###9. Thanks
Sincere thanks are presented to Dr. Matthew D. Green and Meredith L. Patterson, who gave feedback on an early draft of this document.

Sincere thanks are presented to Trevor Perrin for his invaluable contribution to miniLock's design, which introduced sender ID anonymity in the ciphertext.

Sincere thanks are presented to Dmitry Chestnykh for his work on porting TweetNaCL to JavaScript and his general cooperation with the miniLock project, including many helpful and crucial suggestions.

Sincere thanks are presented to Dr. Mario Heiderich and his team at [Cure53](https://cure53.de/) for their work on performing a full audit of the miniLock codebase. We also sincerely thank the [Open Technology Fund](https://www.opentechfund.org/) for funding the audit.

Finally, sincere thanks are presented to the wonderful, constructive members of the miniLock community who have contributed many improvements and ideas to the miniLock design and codebase. You rock!

###10. Credits
####miniLock
Copyright 2014 [Nadim Kobeissi](http://nadim.computer). Released under the AGPLv3 license.

####TweetNaCL
[Daniel J. Bernstein](http://cr.yp.to/djb.html), Wesley Janssen, [Tanja Lange](http://hyperelliptic.org/tanja), [Peter Schwabe](http://www.cryptojedi.org/users/peter/), [Matthew Dempsky](https://github.com/mdempsky)

####TweetNaCL-JS
[Dmitry Chestnykh](http://github.com/dchest), [Devi Mandiri](https://github.com/devi)

####scrypt
[Colin Percival](http://www.tarsnap.com/scrypt.html)


CHANGELOG.md
#Changelog

##miniLock 1.0.2
**August 18 2014**
- miniLock now supports folder encryption. Simply drag and drop the folder into miniLock to compress and encrypt it.
- Bug fixes and improvements.

##miniLock 1.0.1
**August 7 2014**

- Fixed a bug that prevented the encryption of files that certain filenames with international characters.
- Fixed a variety of small user interface inconsistencies, such as text not being centred properly, etc.

##miniLock 1.0
**August 4 2014**

- Initial release.

##miniLock 0.0.1
**July 19 2014**

- Pre-release (Github only) for review purposes.


test/tests/deriveKey.js
// Key derivation test.
QUnit.asyncTest('deriveKey', function(assert) {
	'use strict';
	var passphrase = 'This passphrase is supposed to be good enough for miniLock. :-)'
	miniLock.crypto.getKeyPair(passphrase, 'miniLockScrypt..')
	assert.deepEqual(miniLock.session.keyPairReady, false, 'keyPairReady starts as false')
	assert.deepEqual(Object.keys(miniLock.session.keys).length, 0, 'sessionKeys is empty')
	var keyInterval = setInterval(function() {
		if (miniLock.session.keyPairReady) {
			clearInterval(keyInterval)
			assert.deepEqual(Object.keys(miniLock.session.keys).length, 2, 'sessionKeys is filled')
			assert.deepEqual(miniLock.session.keyPairReady, true, 'keyPairReady set to true')
			assert.deepEqual(typeof(miniLock.session.keys), 'object', 'Type check')
			assert.deepEqual(typeof(miniLock.session.keys.publicKey), 'object', 'Public key type check')
			assert.deepEqual(typeof(miniLock.session.keys.secretKey), 'object', 'Secret key type check')
			assert.deepEqual(miniLock.session.keys.publicKey.length, 32, 'Public key length')
			assert.deepEqual(miniLock.session.keys.secretKey.length, 32, 'Secret key length')

			assert.deepEqual(
				Base58.encode(miniLock.session.keys.publicKey),
				'F9NgHy5pDe5RuvVGe2gRJoPrJ99m1wdE5gutYkVXgjDP',
				'Public key Base58 representation'
			)
			assert.deepEqual(
				nacl.util.encodeBase64(miniLock.session.keys.secretKey),
				'llu8PDzh8JFJbVmwlS90MpubvLogFbuNXqy9fHk3lYw=',
				'Secret key Base64 representation'
			)
			assert.deepEqual(
				miniLock.crypto.getMiniLockID(miniLock.session.keys.publicKey),
				'F9NgHy5pDe5RuvVGe2gRJoPrJ99m1wdE5gutYkVXgjDP',
				'miniLock ID from public key'
			)
			QUnit.start()
		}
	}, 500)
})


test/tests/validateID.js
// Tests for detecting valid miniLock IDs.
QUnit.test('validateID', function(assert) {
	'use strict';
	var IDs = [
		Base58.encode((nacl.box.keyPair()).publicKey),
		Base58.encode((nacl.box.keyPair()).publicKey),
		Base58.encode((nacl.box.keyPair()).publicKey),
		Base58.encode((nacl.box.keyPair()).publicKey)
	]
	var notIDs = [
		'clearly not an ID',
		'ejSSnXzCP806SWiDgeueYlwf2U8utLSkNhwU1VoiAE=',
		'7m+Saj2zhy0mGNKEJPI8V1kyFfZZfcxbWxCcYlwPF0=',
		'4153+icvEiQuiBpttMfHjUsMuy2vDCylzpnTSMQFY2SQ=',
		'u1F4OzpS9PO3ZQitXFDDwTdLsZmLlA3rCNwCYsnBjY=',
		'YWFhYWFhYWFhYWJiYmJiYmJiYmJjY2NjY2NjY2Nj'
	]
	for (var i = 0; i < IDs.length; i++) {
		assert.ok(miniLock.util.validateID(IDs[i]), 'Valid ID ' + i)
	}
	for (var o = 0; o < notIDs.length; o++) {
		assert.ok(!miniLock.util.validateID(notIDs[o]), 'Invalid ID ' + o)
	}
})


src/js/templates.js
miniLock.templates = {}

miniLock.templates.keyStrengthMoreInfo = ''
	+ 'The passphrase you have entered is too weak.'
	+ '<p>Having a long, unique key is very important for using miniLock. '
	+ 'Try using a <strong>phrase</strong> that makes sense only to you.<br /><br />'
	+ 'We strongly recommend allowing miniLock to generate a key for you:<br />'
	+ '<input type="text" value="{{phrase}}" spellcheck="false" readonly="readonly" />'
	+ '<input type="button" value="Get another phrase" /></p>'

miniLock.templates.recipientListIdentity = ''
	+ '<div class="{{className}} identity">'
	+ 	'<label>{{label}}</label>'
	+ 	'<input class="code" type="text" placeholder="Add another miniLock ID" value="{{id}}" spellcheck="false" />'
	+ 	'<input class="remove" type="button" value="&#xf081;" tabindex="-1" />'
	+ '</div>'

miniLock.templates.filename = ''
	+ '<b class="basename">{{basename}}</b>'
	+ '<span class="spaceToAllowWordWrapOnExtensions"> </span>'
	+ '<span class="extensions">{{extensions}}</span>'
	+ '<span class="spaceToAllowWordWrapOnExtensions"> </span>'
	+ '<span class="miniLockExtension">.minilock</span>'


src/js/workers/crypto.js
// -----------------------
// Initialization
// -----------------------
'use strict';

/*jshint -W079 */
var window = {}
/*jshint +W079 */
importScripts(
	'../lib/crypto/nacl.js'
)
var nacl = window.nacl
importScripts(
	'../lib/crypto/nacl-stream.js',
	'../lib/crypto/blake2s.js',
	'../lib/base58.js'
)

// Chunk size (in bytes)
// Warning: Must not be less than 256 bytes
var chunkSize = 1024 * 1024 * 1

// -----------------------
// Utility functions
// -----------------------

var base64Match = new RegExp(
	'^(?:[A-Za-z0-9+\/]{4})*(?:[A-Za-z0-9+\/]{2}==|[A-Za-z0-9+\/]{3}=)?$'
)

var base58Match = new RegExp(
	'^[1-9ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz]+$'
)

// Input: String
// Output: Boolean
// Notes: Validates if string is a proper miniLock ID.
var validateID = function(id) {
	if (
		(id.length > 55) ||
		(id.length < 40)
	) {
		return false
	}
	if (!base58Match.test(id)) {
		return false
	}
	var bytes = Base58.decode(id)
	if (bytes.length !== 33) {
		return false
	}
	var hash = new BLAKE2s(1)
	hash.update(bytes.subarray(0, 32))
	if (hash.digest()[0] !== bytes[32]) {
		return false
	}
	return true
}

// Input: Nonce (Base64) (String), Expected nonce length in bytes (Number)
// Output: Boolean
// Notes: Validates if string is a proper nonce.
var validateNonce = function(nonce, expectedLength) {
	if (
		(nonce.length > 40) ||
		(nonce.length < 10)
	) {
		return false
	}
	if (base64Match.test(nonce)) {
		var bytes = nacl.util.decodeBase64(nonce)
		return bytes.length === expectedLength
	}
	return false
}

// Input: String
// Output: Boolean
// Notes: Validates if string is a proper symmetric key.
var validateKey = function(key) {
	if (
		(key.length > 50) ||
		(key.length < 40)
	) {
		return false
	}
	if (base64Match.test(key)) {
		var bytes = nacl.util.decodeBase64(key)
		return bytes.length === 32
	}
	return false
}

var validateEphemeral = validateKey

// Input: Number
// Output: Number as 4-byte Uint8Array
var numberToByteArray = function(n) {
    var byteArray = [0, 0, 0, 0]
    for (var i = byteArray.length - 1; i >= 0; i--) {
        byteArray[i] = n & 255
		n = n >> 8
    }
    return new Uint8Array(byteArray)
}

// Input: 4-byte Uint8Array
// Output: ByteArray converter to number
var byteArrayToNumber = function(byteArray) {
	var n = 0
	for (var i = 0; i < byteArray.length; i++) {
		n += byteArray[i]
		if (i < byteArray.length-1) {
			n = n << 8
		}
	}
	return n
}

// -----------------------
// Cryptographic functions
// -----------------------

// Receive a message to perform a certain operation.
// Input: Object:
//	{
//		operation: Type of operation ('encrypt' or 'decrypt'),
//		data: Data to encrypt/decrypt (Uint8Array),
//		name: File name (String),
//		saveName: Name to use for saving resulting file (String),
//		fileKey: 32-byte key used for file encryption (Uint8Array),
//		fileNonce: 16-byte nonce used for file encryption/decryption (Uint8Array),
//		decryptInfoNonces: Array of 24-byte nonces (Uint8Array) to be used to encrypt
//			decryptInfo and fileInfo objects (one for each recipient)
//		ephemeral: {
//			publicKey: Ephemeral Curve25519 public key (Uint8Array),
//			secretKey: Ephemeral Curve25519 secret key (Uint8Array)
//		} (Only used for encryption)
//		miniLockIDs: Array of (Base58) miniLock IDs to encrypt to (not used for 'decrypt' operation),
//		myMiniLockID: Sender's miniLock ID (String),
//		mySecretKey: Sender's secret key (Uint8Array)
//	}
// Result: When finished, the worker will return the result
// 	which is supposed to be caught and processed by
//	the miniLock.crypto.worker.onmessage() function
//	in miniLock.js.
// Notes: A miniLock-encrypted file's first 8 bytes are always the following:
//	0x6d, 0x69, 0x6e, 0x69,
//	0x4c, 0x6f, 0x63, 0x6b,
//	Those 8 bytes are then followed by an 4-byte little-endian value indicating the byte length
//	of the file header, which is the following JSON object (binary-encoded):
//	{
//		version: Version of the miniLock protocol used for this file (Currently 1) (Number)
//		ephemeral: Public key from ephemeral key pair used to encrypt decryptInfo object (Base64),
//		decryptInfo: {
//			(One copy of the below object for every recipient)
//			Unique nonce for decrypting this object (Base64): {
//				senderID: Sender's miniLock ID (Base58),
//				recipientID: miniLock ID of this recipient (used for verfication) (Base58),
//				fileInfo: {
//					fileKey: Key for file decryption (Base64),
//					fileNonce: Nonce for file decryption (Base64),
//					fileHash: BLAKE2 hash (32 bytes) of the ciphertext bytes. (Base64)
//				} (fileInfo is encrypted to recipient's public key using long-term key pair) (Base64),
//			} (encrypted to recipient's public key using ephemeral key pair) (Base64)
//
//		}
//	}
//	The file's filename is padded with 0x00 bytes until its length equals 256 bytes.
//	The filename is then prepended to the plaintext prior to encryption and is encrypted as its own chunk.
//	Note that the nonce used to encrypt decryptInfo is the same as the one used to encrypt fileInfo.
//	Nonce reuse in this scenario is permitted since we are encrypting using different keys.
//	The ciphertext in binary format is appended after the header.
onmessage = function(message) {
message = message.data

// We have received a request to encrypt
if (message.operation === 'encrypt') {
	(function() {
		var header = {
			version: 1,
			ephemeral: nacl.util.encodeBase64(message.ephemeral.publicKey),
			decryptInfo: {}
		}
		var streamEncryptor = nacl.stream.createEncryptor(
			message.fileKey,
			message.fileNonce,
			chunkSize
		)
		var paddedFileName = message.name
		while (paddedFileName.length < 256) {
			paddedFileName += String.fromCharCode(0x00)
		}
		var fileHash = new BLAKE2s(32)
		var encrypted = []
		var encryptedChunk
		encryptedChunk = streamEncryptor.encryptChunk(
			nacl.util.decodeUTF8(paddedFileName),
			false
		)
		fileHash.update(encryptedChunk)
		encrypted.push(encryptedChunk)
		for (var c = 0; c < message.data.length; c += chunkSize) {
			if (c >= (message.data.length - chunkSize)) {
				encryptedChunk = streamEncryptor.encryptChunk(
					message.data.subarray(c),
					true
				)
			}
			else {
				encryptedChunk = streamEncryptor.encryptChunk(
					message.data.subarray(c, c + chunkSize),
					false
				)
			}
			if (!encryptedChunk) {
				postMessage({
					operation: 'encrypt',
					error: 1
				})
				throw new Error('miniLock: Encryption failed - general encryption error')
				return false
			}
			fileHash.update(encryptedChunk)
			encrypted.push(encryptedChunk)
			postMessage({
				operation: 'encrypt',
				progress: c + chunkSize,
				total: message.data.length
			})
		}
		streamEncryptor.clean()
		for (var i = 0; i < message.miniLockIDs.length; i++) {
			var decryptInfo = {
				senderID: message.myMiniLockID,
				recipientID: message.miniLockIDs[i],
				fileInfo: {
					fileKey: nacl.util.encodeBase64(message.fileKey),
					fileNonce: nacl.util.encodeBase64(message.fileNonce),
					fileHash: nacl.util.encodeBase64(fileHash.digest())
				}
			}
			decryptInfo.fileInfo = nacl.util.encodeBase64(nacl.box(
				nacl.util.decodeUTF8(JSON.stringify(decryptInfo.fileInfo)),
				message.decryptInfoNonces[i],
				Base58.decode(message.miniLockIDs[i]).subarray(0, 32),
				message.mySecretKey
			))
			decryptInfo = nacl.util.encodeBase64(nacl.box(
				nacl.util.decodeUTF8(JSON.stringify(decryptInfo)),
				message.decryptInfoNonces[i],
				Base58.decode(message.miniLockIDs[i]).subarray(0, 32),
				message.ephemeral.secretKey
			))
			header.decryptInfo[
				nacl.util.encodeBase64(message.decryptInfoNonces[i])
			] = decryptInfo
		}
		header = JSON.stringify(header)
		encrypted.unshift(
			'miniLock',
			numberToByteArray(header.length),
			header
		)
		encrypted = (new FileReaderSync()).readAsArrayBuffer(new Blob(encrypted))
		postMessage({
			operation: 'encrypt',
			blob: encrypted,
			name: message.name,
			saveName: message.saveName,
			senderID: message.myMiniLockID,
			callback: message.callback
		}, [encrypted])
		encrypted = null
	})()
}


// We have received a request to decrypt
if (message.operation === 'decrypt') {
	(function() {
		var header, headerLength
		try {
			headerLength = byteArrayToNumber(
				new Uint8Array(message.data.subarray(8, 12))
			)
			header = nacl.util.encodeUTF8(
				message.data.subarray(12, headerLength + 12)
			)
			header = JSON.parse(header)
			message.data = message.data.subarray(
				12 + headerLength,
				message.data.length
			)
		}
		catch(error) {
			postMessage({
				operation: 'decrypt',
				error: 3
			})
			throw new Error('miniLock: Decryption failed - could not parse header')
			return false
		}
		if (
			!header.hasOwnProperty('version')
			|| header.version !== 1
		) {
			postMessage({
				operation: 'decrypt',
				error: 4
			})
			throw new Error('miniLock: Decryption failed - invalid header version')
			return false
		}
		if (
			!header.hasOwnProperty('ephemeral')
			|| !validateEphemeral(header.ephemeral)
		) {
			postMessage({
				operation: 'decrypt',
				error: 5
			})
			throw new Error('miniLock: Decryption failed - could not validate sender ID')
			return false
		}
		// Attempt decryptInfo decryptions until one succeeds
		var actualDecryptInfo      = null
		var actualDecryptInfoNonce = null
		var actualFileInfo         = null
		for (var i in header.decryptInfo) {
			if (
				({}).hasOwnProperty.call(header.decryptInfo, i)
				&& validateNonce(i, 24)
			) {
				try {
					nacl.util.decodeBase64(header.decryptInfo[i])
				}
				catch(err) {
					postMessage({
						operation: 'decrypt',
						error: 3
					})
					throw new Error('miniLock: Decryption failed - could not parse header')
					return false
				}
				actualDecryptInfo = nacl.box.open(
					nacl.util.decodeBase64(header.decryptInfo[i]),
					nacl.util.decodeBase64(i),
					nacl.util.decodeBase64(header.ephemeral),
					message.mySecretKey
				)
				if (actualDecryptInfo) {
					try {
						actualDecryptInfo = JSON.parse(
							nacl.util.encodeUTF8(actualDecryptInfo)
						)
						actualDecryptInfoNonce = nacl.util.decodeBase64(i)
					}
					catch(err) {
						postMessage({
							operation: 'decrypt',
							error: 3
						})
						throw new Error('miniLock: Decryption failed - could not parse header')
						return false
					}
					break
				}
			}
		}
		if (
			!actualDecryptInfo
			|| !({}).hasOwnProperty.call(actualDecryptInfo, 'recipientID')
			|| actualDecryptInfo.recipientID !== message.myMiniLockID
		) {
			postMessage({
				operation: 'decrypt',
				error: 6
			})
			throw new Error('miniLock: Decryption failed - File is not encrypted for this recipient')
			return false
		}
		if (
			!({}).hasOwnProperty.call(actualDecryptInfo, 'fileInfo')
			|| !({}).hasOwnProperty.call(actualDecryptInfo, 'senderID')
			|| !validateID(actualDecryptInfo.senderID)
		) {
			postMessage({
				operation: 'decrypt',
				error: 3
			})
			throw new Error('miniLock: Decryption failed - could not parse header')
			return false
		}
		try {
			actualFileInfo = nacl.box.open(
				nacl.util.decodeBase64(actualDecryptInfo.fileInfo),
				actualDecryptInfoNonce,
				Base58.decode(actualDecryptInfo.senderID).subarray(0, 32),
				message.mySecretKey
			)
			actualFileInfo = JSON.parse(
				nacl.util.encodeUTF8(actualFileInfo)
			)
			actualFileInfo.fileKey   = nacl.util.decodeBase64(actualFileInfo.fileKey)
			actualFileInfo.fileNonce = nacl.util.decodeBase64(actualFileInfo.fileNonce)
			actualFileInfo.fileHash  = nacl.util.decodeBase64(actualFileInfo.fileHash)
		}
		catch(err) {
			postMessage({
				operation: 'decrypt',
				error: 3
			})
			throw new Error('miniLock: Decryption failed - could not parse header')
			return false
		}
		var streamDecryptor = nacl.stream.createDecryptor(
			actualFileInfo.fileKey,
			actualFileInfo.fileNonce,
			chunkSize
		)
		var fileHash = new BLAKE2s(32)
		var decrypted = []
		var chunk, decryptedChunk
		chunk = message.data.subarray(0, 4 + 16 + 256)
		decryptedChunk = streamDecryptor.decryptChunk(
			chunk,
			false
		)
		fileHash.update(chunk)
		decrypted.push(decryptedChunk)
		for (var c = 4 + 16 + 256; c < message.data.length; c += (4 + 16 + chunkSize)) {
			if (c >= (message.data.length - (4 + 16 + chunkSize))) {
				chunk = message.data.subarray(c)
				decryptedChunk = streamDecryptor.decryptChunk(
					chunk,
					true
				)
			}
			else {
				chunk = message.data.subarray(c, c + (4 + 16 + chunkSize))
				decryptedChunk = streamDecryptor.decryptChunk(
					chunk,
					false
				)
			}
			if (!decryptedChunk) {
				postMessage({
					operation: 'decrypt',
					error: 2
				})
				throw new Error('miniLock: Decryption failed - general decryption error')
				return false
			}
			fileHash.update(chunk)
			decrypted.push(decryptedChunk)
			postMessage({
				operation: 'decrypt',
				progress: c + chunkSize,
				total: message.data.length
			})
		}
		streamDecryptor.clean()
		var fileName = ''
		try {
			fileName = nacl.util.encodeUTF8(decrypted[0])
			decrypted.splice(0, 1)
			while (
				fileName[fileName.length - 1] === String.fromCharCode(0x00)
			) {
				fileName = fileName.slice(0, -1)
			}
			decrypted = (new FileReaderSync()).readAsArrayBuffer(
				new Blob(decrypted)
			)
		}
		catch(err) {
			postMessage({
				operation: 'decrypt',
				error: 2
			})
			throw new Error('miniLock: Decryption failed - general decryption error')
			return false
		}
		if (
			!nacl.verify(
				new Uint8Array(fileHash.digest()),
				actualFileInfo.fileHash
			)
		) {
			postMessage({
				operation: 'decrypt',
				error: 7
			})
			throw new Error('miniLock: Decryption failed - could not validate file contents after decryption')
			return false
		}
		postMessage({
			operation: 'decrypt',
			blob: decrypted,
			name: fileName,
			saveName: fileName,
			senderID: actualDecryptInfo.senderID,
			callback: message.callback
		}, [decrypted])
		decrypted = null
	})()
}

}
